hack/generator/pkg/functions/property_assignment_function.go

```diff
@@ -331,3 +381,76 @@ func (fn *PropertyAssignmentFunction) createConversion(
 		return conversion(reader, writer, generationContext)
 	}, nil
 }
+
+// findPropertyBagProperty looks for a property bag on the specified type and returns it if found, or nil otherwise
+// We recognize the property bag by type, so that the name can vary to avoid collisions with other properties if needed.
+func (fn *PropertyAssignmentFunction) findPropertyBagProperty(instance astmodel.Type) (*astmodel.PropertyDefinition, bool) {
+	if container, ok := astmodel.AsPropertyContainer(instance); ok {
+		for _, prop := range container.Properties() {
+			if prop.PropertyType().Equals(astmodel.PropertyBagType) {
+				return prop, true
+			}
+		}
+	}
+
+	return nil, false
+}
+
+// sourceType returns the type we are reading information from
+// When converting FROM, otherDefinition.Type() is our source
+// When converting TO, receiverDefinition.Type() is our source
+// Our inverse is destinationType()
+func (fn *PropertyAssignmentFunction) sourceType() astmodel.Type {
+	return fn.direction.SelectType(fn.otherDefinition.Type(), fn.receiverDefinition.Type())
+}
+
+// destinationType returns the type we are writing information from
+// When converting FROM, receiverDefinition.Type() is our source
+// When converting TO, otherDefinition.Type() is our source
+// Our inverse is sourceType()
+func (fn *PropertyAssignmentFunction) destinationType() astmodel.Type {
+	return fn.direction.SelectType(fn.receiverDefinition.Type(), fn.otherDefinition.Type())
+}
+
+// createReadingEndpoints creates a ReadableConversionEndpointSet containing all the readable endpoints we need for this
+// conversion. If the source has a property bag, we create additional endpoints to match any surplus properties present
+// on the DESTINATION type, so we can populate those from the property bag. Returns true if we create any endpoints to
+// read from a property bag, false otherwise.
+func (fn *PropertyAssignmentFunction) createReadingEndpoints() (conversions.ReadableConversionEndpointSet, bool) {
+	sourceEndpoints := conversions.NewReadableConversionEndpointSet()
+	sourceEndpoints.CreatePropertyEndpoints(fn.sourceType(), fn.knownLocals)
+	sourceEndpoints.CreateValueFunctionEndpoints(fn.sourceType(), fn.knownLocals)
+
+	readsFromPropertyBag := false
+	if _, found := fn.findPropertyBagProperty(fn.sourceType()); found {
+		// Add source endpoints for DESTINATION properties that don't already have matching source properties
+		count := sourceEndpoints.CreateBagItemEndpoints(fn.destinationType(), fn.knownLocals)
```

Comment:
Yeah this is tricksy - I think it would warrant a comment on `CreateBagItemEndpoints` pointing at this behaviour just for extra clarity.

Reply:
I've improved the comments on `CreateBagItemEntpoints`, and have expanded on the comment here as well.